File: src/project_manager_flutter/lib/features/time_entries/repository/time_entry_repository.dart
import 'package:dio/dio.dart';
import '../../../core/constants/api_constants.dart';
import '../../../core/network/api_exceptions.dart';
import '../models/create_time_entry_request.dart';
import '../models/paginated_time_entries.dart';
import '../models/time_entry_detail.dart';
import '../models/update_time_entry_request.dart';

class TimeEntryRepository {
  final Dio _dio;

  TimeEntryRepository(this._dio);

  Future<PaginatedTimeEntries> listTimeEntries({
    String? projectId,
    String? userId,
    String? fromDate,
    String? toDate,
    String? taskId,
    int page = 1,
    int pageSize = 20,
  }) async {
    try {
      final response = await _dio.get(
        ApiConstants.timeEntries,
        queryParameters: {
          'page': page,
          'pageSize': pageSize,
          if (projectId != null) 'projectId': projectId,
          if (userId != null) 'userId': userId,
          if (fromDate != null) 'fromDate': fromDate,
          if (toDate != null) 'toDate': toDate,
          if (taskId != null) 'taskId': taskId,
        },
      );
      return PaginatedTimeEntries.fromJson(response.data);
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<TimeEntryDetail> getTimeEntry(String id) async {
    try {
      final response = await _dio.get(ApiConstants.timeEntryById(id));
      return TimeEntryDetail.fromJson(response.data);
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<void> createTimeEntry(CreateTimeEntryRequest request) async {
    try {
      await _dio.post(ApiConstants.timeEntries, data: request.toJson());
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<void> updateTimeEntry(String id, UpdateTimeEntryRequest request) async {
    try {
      await _dio.put(ApiConstants.timeEntryById(id), data: request.toJson());
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<void> deleteTimeEntry(String id) async {
    try {
      await _dio.delete(ApiConstants.timeEntryById(id));
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }
}

File: src/project-manager-web/src/features/time-entries/repository/time-entry-repository.ts
import { apiClient } from "@/core/api/client";
import { ApiEndpoints } from "@/core/api/constants";
import { ApiException } from "@/core/api/exceptions";
import type { TimeEntry, CreateTimeEntryRequest, UpdateTimeEntryRequest } from "../models/types";
import type { PaginatedResponse } from "@/core/api/types";
import { AxiosError } from "axios";

function handleError(error: unknown): never {
  if (error instanceof AxiosError) {
    throw ApiException.fromResponse(error.response?.status ?? 500, error.response?.data);
  }
  throw error;
}

export interface TimeEntryListFilters {
  projectId?: string;
  taskId?: string;
  userId?: string;
  clientId?: string;
  fromDate?: string;
  toDate?: string;
  unpaged?: boolean;
  pageSize?: number;
}

export const timeEntryRepository = {
  async list(filters?: TimeEntryListFilters): Promise<TimeEntry[]> {
    try {
      const { unpaged, pageSize, ...rest } = filters ?? {};
      const params: Record<string, unknown> = { ...rest };
      if (unpaged) {
        params.unpaged = true;
      } else {
        params.pageSize = pageSize ?? 100;
      }
      const response = await apiClient.get<PaginatedResponse<TimeEntry>>(ApiEndpoints.timeEntries, { params });
      return response.data.items;
    } catch (error) { handleError(error); }
  },

  async get(id: string): Promise<TimeEntry> {
    try {
      const response = await apiClient.get<TimeEntry>(ApiEndpoints.timeEntryById(id));
      return response.data;
    } catch (error) { handleError(error); }
  },

  async create(request: CreateTimeEntryRequest): Promise<void> {
    try {
      await apiClient.post(ApiEndpoints.timeEntries, request);
    } catch (error) { handleError(error); }
  },

  async update(id: string, request: UpdateTimeEntryRequest): Promise<void> {
    try {
      await apiClient.put(ApiEndpoints.timeEntryById(id), request);
    } catch (error) { handleError(error); }
  },

  async delete(id: string): Promise<void> {
    try {
      await apiClient.delete(ApiEndpoints.timeEntryById(id));
    } catch (error) { handleError(error); }
  },
};
